lib/explorer/data_frame.ex

```diff
@@ -521,28 +529,80 @@ defmodule Explorer.DataFrame do
 
     * `:columns` - A list of column names or indexes to keep. If present,
       only these columns are read into the dataframe. (default: `nil`)
+
+    * `:config` - An optional config struct, normally associated with the
+      filesystem entry provided. For example, if an S3 object is being fetch,
+      then config is going to be something like `%Filesystem.S3.Config{}`.
+      Usage: `from_parquet("s3://bkt/file", config: %Filesystem.S3.Config{}`).
   """
   @doc type: :io
-  @spec from_parquet(filename :: String.t(), opts :: Keyword.t()) ::
+  @spec from_parquet(entry :: String.t() | fs_entry(), opts :: Keyword.t()) ::
           {:ok, DataFrame.t()} | {:error, term()}
-  def from_parquet(filename, opts \\ []) do
+  def from_parquet(entry, opts \\ []) do
     {backend_opts, opts} = Keyword.split(opts, [:backend, :lazy])
 
     opts =
       Keyword.validate!(opts,
         max_rows: nil,
-        columns: nil
+        columns: nil,
+        config: nil
       )
 
     backend = backend_from_options!(backend_opts)
 
     backend.from_parquet(
-      filename,
+      normalise_entry(entry, opts[:config]),
       opts[:max_rows],
       to_columns_for_io(opts[:columns])
     )
   end
 
+  defp normalise_entry(%Local.Entry{} = entry, _config), do: entry
+  defp normalise_entry(%S3.Entry{config: %S3.Config{}} = entry, nil), do: entry
+  defp normalise_entry(%S3.Entry{config: %S3.Config{} = config} = entry, config), do: entry
+
+  defp normalise_entry(%S3.Entry{config: nil} = entry, %S3.Config{} = config),
+    do: %{entry | config: config}
+
+  # should we allow overwrites?
+  defp normalise_entry(%S3.Entry{}, config) do
+    raise "incompatible entry configuration with provided config: #{inspect(config)}."
```

Comment:
```suggestion
    raise ArgumentError, "incompatible entry configuration with provided config: #{inspect(config)}"
```

Reply:
I think you can have this at the top:

    defp normalise_entry(%_{} = entry, config) when config != nil do
      raise ArgumentError, ":config key is only supported when the argument is a string, got #{inspect(entry)} with config #{inspect(config)}"
    end

And then we can assume all entries have no config forward.